### ISSUE TITLE
Special characters like Ä, Ü and Ö cant be identified by fast alpr unfortunately. Is there a solution?

### ISSUE THREAD

**Issue by xXHTXx** [ankandrew/fast-alpr#19]
First of all, I would like to sincerely thank you for your outstanding work on this open-source project – it has been incredibly helpful, and I truly appreciate the time and effort you’ve invested in it.

As I live in Germany, many license plates here contain umlauts such as Ä, Ü, and Ö. Unfortunately, Fast ALPR doesn’t seem to recognize these special characters.

If there is any known workaround or future update planned to support these characters, I would be very grateful for any information or advice you could share.

Thank you once again for your time and support!

**Comment by ankandrew** [comment]
Hi! Great you are find useful `fast-alpr`! Yep, it's expected that umlauts are not recognized because I normalized them (i.e. Ä -> A). I can train another model to have support for those characters. Can you give me a list of countries that include that sort of characters? Does any other European country have those characters?

**Comment by xXHTXx** [comment]
I truly appreciate you taking the time to make that possible. I hope it won’t be too much of a hassle, thank you so much!
As far as I know only Germany systematically includes Ä, Ö, Ü in standard license plates. A few others (Austria, Switzerland, Liechtenstein) permit them in special cases, but no other European country has them in their default registration formats.

**Comment by ankandrew** [comment]
Digging a bit more, I'm confused about the structure of German plates. Aren't the umlauts only present in the location part of the plate (first part from left-to-right)?

Also from [wikipedia](https://en.wikipedia.org/wiki/Vehicle_registration_plates_of_Germany):

> Germany includes diacritical marks in the letters of some codes, that is the letters Ä,[[14]](https://en.wikipedia.org/wiki/Vehicle_registration_plates_of_Germany#cite_note-26) Ö, and Ü,[[i]](https://en.wikipedia.org/wiki/Vehicle_registration_plates_of_Germany#cite_note-27) but not [ẞ](https://en.wikipedia.org/wiki/%C3%9F).[[j]](https://en.wikipedia.org/wiki/Vehicle_registration_plates_of_Germany#cite_note-28) For a long time, German codes kept to the rule that a code with an umlaut would prohibit another code with the respective blank vowel, e.g. there could not be a district code FU as the code FÜ was already in use for [Fürth](https://en.wikipedia.org/wiki/F%C3%BCrth). This rule was disregarded in 1996, when BÖ was introduced for [Bördekreis](https://en.wikipedia.org/wiki/B%C3%B6rdekreis) in spite of BO existing for [Bochum](https://en.wikipedia.org/wiki/Bochum). However Bördekreis got BK in 2007, probably because of confusions abroad, for fines etc.[[k]](https://en.wikipedia.org/wiki/Vehicle_registration_plates_of_Germany#cite_note-29)

Nowadays can there be different plates for i.e. `LöXXX` and another for `LOXXX`? I'm asking because googling plates I always see at the start of the plate these umlaut chars.

**Comment by xXHTXx** [comment]
You're right to be confused, German license plates can indeed be a bit tricky when it comes to Umlaute! But in general:

Umlaute are (almost) always in the location code (first part) as you mentioned. The first part (before the sticker/tag) is the Unterscheidungszeichen (location code), which represents the city/district. This is where Umlaute (Ä, Ö, Ü) appear, e.g., LÜ (Lüneburg), HÖ (Höxter), BÜ (Bütthard).

**Comment by ankandrew** [comment]
So in practice, we won't have Unterscheidungszeichen/location codes that after normalizing the Umlaute they can be confused with another city/district? If that is the case couldn't you do a hand-rule to perform the char replacement? I'm saying this because this part of the Wikipedia quote:

> ... e.g. there could not be a district code FU as the code FÜ was already in use for Fürth. This rule was disregarded in 1996, when BÖ was introduced for Bördekreis in spite of BO existing for Bochum. However Bördekreis got BK in 2007, probably because of confusions abroad, for fines etc.

Also from same article says that this chars could also appear in other personalized plates of these countries (Finnish and Swedish):

> This habit is uncommon in most other European countries but does appear on regular [Montenegrin](https://en.wikipedia.org/wiki/Vehicle_registration_plates_of_Montenegro), [Croatian](https://en.wikipedia.org/wiki/Vehicle_registration_plates_of_Croatia) and [Serbian](https://en.wikipedia.org/wiki/Vehicle_registration_plates_of_Serbia) (letters Đ, Č, Ć, Š and Ž), [Åland registration plates](https://en.wikipedia.org/wiki/Vehicle_registration_plates_of_Finland) (letters Å, Ä and Ö), as well as on [Finnish](https://en.wikipedia.org/wiki/Vehicle_registration_plates_of_Finland), [Swedish](https://en.wikipedia.org/wiki/Vehicle_registration_plates_of_Sweden#Personal_plates) (letters Å, Ä and Ö) and [Danish](https://en.wikipedia.org/wiki/Vehicle_registration_plates_of_Denmark) (letters Æ, Ø and Å) personal registration plates.

**Comment by xXHTXx** [comment]
In practice, yes. After normalizing umlauts, German district codes should theoretically avoid conflicts. But in 1996, they broke this rule with BÖ (which clashed with BO). They fixed it in 2007 by changing BÖ to BK after it caused problems abroad. The example you cited shows this wasn't always enforced historically

Other countries (like Sweden or Croatia) allow special letters (Å, Ä, Đ, Č, etc.) on custom vanity plates. But Germany's codes are official – so they can't afford mistakes.

If it’s not too much trouble, I’d be super happy if you could add support for detecting German umlauts. I’ve worked with PaddleOCR and EasyOCR before in combination with YOLOv10 and also tried OpenALPR, but none of them gave results as good as what you’ve built here. Huge respect for that!

**Comment by ankandrew** [comment]
Great, makes sense! Thanks for the info. I was just checking the dataset I have and I believe it's not sufficient, I checked for plates that have any of the umlauts and are only 113 out of the total 4988 german plates I have (2.2% approx.) I think this is a low number. You can either collect data or more easily just normalize with hand-rules (given that there should be little conflicts).

**Comment by ankandrew** [comment]
I will soon release new models that converge better (also trained with big dataset) and write a officual tutorial to fine-tune the model. Stay tuned!!

**Comment by xXHTXx** [comment]
I truly appreciate your help, you’re amazing. If it’s helpful, I can gather some data that includes license plates with umlauts.

**Comment by ankandrew** [comment]
Cool! If you can collect plates w/ umlauts it would be great. We can train a model once we have a little more data. If we could collect 400 with umaults it would be nice.

**Comment by xXHTXx** [comment]
I'm currently working on collecting around 400–500 images for each license plate that contains the characters Ä, Ü, and Ö.
I plan to structure everything with one main folder called "Images" and then create subfolders for Ä, Ü, and Ö, each containing the 400–500 images. Would that be enough for your purposes? It might take me a few days to complete everything. Thank you for your understanding and your time!

**Comment by ankandrew** [comment]
Hi! If you can follow the format expected by [fast-plate-ocr](https://github.com/ankandrew/fast-plate-ocr#:~:text=A%20labeled%20dataset%2C%20see%20arg_plate_dataset.zip%20for%20the%20expected%20data%20format.) (see https://github.com/ankandrew/fast-plate-ocr/releases/download/arg-plates/arg_plate_dataset.zip) it would be much easier for me, so I can train it directly :).

**Comment by xXHTXx** [comment]
Good that you informed me, I had actually read it a few days ago but completely forgot. Alright, I will send you the cropped license plates so that you have them directly, and I will also work on the Excel table. I noticed that in your example, the cropped license plates were in black and white. Should I also send you the cropped license plates in black and white format?

**Comment by ankandrew** [comment]
Don't convert them to black and white, leave them as they are. I'm adding an option in the new release to support RGB images too instead of only grayscale (might be useful in the future when recognizing countries apart from the plate text).

To facilitate the process you can rely on detector so you don't have to crop by hand, like the following:

```py
from open_image_models import LicensePlateDetector
import cv2
from pathlib import Path

img_dir = Path("path/to/folder/with/images")
save_dir = Path("path/to/output/folder")
save_dir.mkdir(parents=True, exist_ok=True)

img_files = (x for x in img_dir.iterdir() if x.suffix in (".jpeg", ".jpg", ".webp", ".png"))

lp_detector = LicensePlateDetector(detection_model="yolo-v9-s-608-license-plate-end2end")

for img_file in img_files:
    img = cv2.imread(str(img_file))
    plates = lp_detector.predict(img)
    for i, plate in enumerate(plates):
        x1, y1, x2, y2 = plate.bounding_box
        plate_crop = img[y1:y2, x1:x2]
        plate_crop_rgb = cv2.cvtColor(plate_crop, cv2.COLOR_BGR2RGB)
        cv2.imwrite(str(save_dir / f"{img_file.stem}_plate_{i}.png"), plate_crop_rgb)
```

_Note: I wrote this quickly just now, make sure it's correct._

Writing this I realized you can also just use `fast-alpr` directly to get the crop + preliminary text (and then you add the umlaut).

**Comment by ankandrew** [comment]
Just noticed your edit, I was going to use part of those 500 for validation. (But the more and diverse plates you can get the better :)).

**Comment by xXHTXx** [comment]
Your suggestion was brilliant, it helped me a lot and worked really well! I have already been able to collect 500 training samples for 'Ü' and 220 validation samples. The CSV for that is already done. I'll continue with 'Ö' and 'Ä' in the coming days. I really appreciate your support! :)

**Comment by xXHTXx** [comment]
I hope you're doing well! 😊
I've uploaded the datasets with the umlauts Ö, Ü, and Ä to my Google Drive, you can download them from there.
https://drive.google.com/file/d/1WJ8fJTgRpV1k3PpLCpA1EjAn_ClK8xoi/view?usp=sharing

Unfortunately, I couldn’t find many images with license plates containing an Ä, since there’s only one city in Germany that has an Ä in its plate code, and that’s Bad Säckingen (SÄK). But I managed to gather quite a few for Ü and Ö.

Feel free to let me know if everything looks good, or if you’d like me to make any adjustments or provide more data.
Thanks a lot!

**Comment by ankandrew** [comment]
Hi, awesome work! I want to soon release the new model architecture and train a model with that data (prob fine-tuned from a new global version). Really nice how you split and put together the data.

I was thinking to include country classification to the model but I will probably do incremental releases (so we can test this new umlaut model before).

Once I have the model it would be great if you can help me verifying how well it works :).  I'll keep up updated.

**Comment by xXHTXx** [comment]
Glad you enjoyed it! Your plan sounds super exciting, especially with the step-by-step release and the focus on the new umlaut model. Adding country classification later makes total sense and sounds like a great next step.

I’d be more than happy to help test the model once it’s ready!

**Comment by ankandrew** [comment]
Hi @xXHTXx! I have the model ready so you can test it. I trained only on german data, combined your dataset with mine to get a total of 8300 images for train and 2500 for validation. It has `95.6`% plate accuracy on the umlaut validation set. This model should be more efficient also to run. It's a modified and smaller architecture based on CCT https://arxiv.org/pdf/2104.05704. It converges much faster than the previous models too.

To run this model do the following:

```py
import requests

from fast_alpr import ALPR

image_url = (
    "https://preview.redd.it/first-time-seeing-an-umlaut-on-a-license-plate-v0-"
    "i9wca6ghy9za1.jpg?width=1080&crop=smart&auto=webp&s=79e09253e08416795d8e98"
    "0e3811764963dc6e6f"
)
test_image_name = "image_name.webp"

img_data = requests.get(image_url).content
with open(test_image_name, "wb") as f_out:
    f_out.write(img_data)

alpr = ALPR(
    detector_model="yolo-v9-s-608-license-plate-end2end",
    ocr_model_path="./2025-05-05_17-42-03/ckpt-epoch_424-acc_0.947.onnx",
    ocr_config_path="./2025-05-05_17-42-03/config.yaml",
)

# Predict
print(alpr.predict(test_image_name))
```

It returns for the image:

```py
[
    ALPRResult(
        detection=DetectionResult(
            label="License Plate",
            confidence=0.90771484375,
            bounding_box=BoundingBox(x1=360, y1=907, x2=663, y2=1005),
        ),
        ocr=OcrResult(text="PLÖMF100", confidence=0.94140625),
    )
]
```

Feel free to test it with diverse data and let me know so we can improve it.

[2025-05-05_17-42-03.zip](https://github.com/user-attachments/files/20049582/2025-05-05_17-42-03.zip)

**Comment by xXHTXx** [comment]
Hi @ankandrew hope you are doing fine.
Thank you so much for the time and effort you put into training and providing the data. I really appreciate your work and dedication! 
I’m honestly super excited about testing it. Even though I’m currently a bit tied up, I’ll make it a priority to give it a proper try as soon as possible. I truly can’t wait to see how it performs!

Also, I’m really impressed by your deep understanding of CCT and how you managed to apply it so effectively in a real project. That kind of skill and creativity is seriously inspiring.

I’ll get back to you with results and impressions as soon as I can!
Take care, my friend.

**Comment by ankandrew** [comment]
@xXHTXx Any news? Some feedback would be nice.

**Comment by xXHTXx** [comment]
@ankandrew 
I am sorry for keeping you waiting for so long. We currently have some construction work going on due to our move, which has caused some delays. I did manage to do a bit of testing earlier, but unfortunately not long enough. Still, I did notice a few things and wrote them down. The construction work should be finished by next week, so I’ll be able to test more thoroughly and I hope I’ll be able to share my full feedback with you by Wednesday. Thanks so much for your patience and understanding!

**Comment by xXHTXx** [comment]
Hi @ankandrew ,
I hope you're doing well! We're almost done with our construction work, and I finally had the chance to properly test everything. It’s working really well — thank you so much! There are no issues at all.

The things I had noted down have all been resolved. For example, I kept getting “PLÖMF” when checking a license plate, but that was my mistake. Everything's working perfectly now, thanks again!

I’d really love to contribute to making fast-alpr even better and stronger for the future. If there’s anything I can do to help, just let me know anytime!
